src/main/scala/io/iohk/ethereum/network/p2p/Message.scala

```diff
@@ -60,44 +64,427 @@ object Hello {
+  override def code: Int = Status.code
+
+  override def toString: String = {
+    s"""Status {
+       |protocolVersion: $protocolVersion
+       |networkId: $networkId
+       |totalDifficulty: $totalDifficulty
+       |bestHash: ${Hex.toHexString(bestHash.toArray[Byte])}
+       |genesisHash: ${Hex.toHexString(genesisHash.toArray[Byte])}
+       |}""".stripMargin
+  }
+}
+
+object NewBlockHashes {
+  implicit val rlpEndDec = new RLPEncoder[NewBlockHashes] with RLPDecoder[NewBlockHashes] {
+    override def encode(obj: NewBlockHashes): RLPEncodeable = {
+      import obj._
+      RLPList(hashes.map(BlockHash.rlpEndDec.encode): _*)
+    }
+
+    override def decode(rlp: RLPEncodeable): NewBlockHashes = rlp match {
+      case rlpList: RLPList => NewBlockHashes(rlpList.items.map(BlockHash.rlpEndDec.decode))
+      case _ => throw new RuntimeException("Cannot decode NewBlockHashes")
+    }
+  }
+
+  val code: Int = 0x10 + 0x01
+}
+
+case class NewBlockHashes(hashes: Seq[BlockHash]) extends Message {
+  override def code: Int = NewBlockHashes.code
+}
+
+
+object BlockHash {
+  implicit val rlpEndDec = new RLPEncoder[BlockHash] with RLPDecoder[BlockHash] {
+    override def encode(obj: BlockHash): RLPEncodeable = {
+      import obj._
+      RLPList(hash.toArray[Byte], number)
+    }
+
+    override def decode(rlp: RLPEncodeable): BlockHash = rlp match {
+      case RLPList(hash, number) => BlockHash(ByteString(hash: Array[Byte]), number)
+      case _ => throw new RuntimeException("Cannot decode BlockHash")
+    }
+  }
+}
+
+case class BlockHash(hash: ByteString, number: BigInt) {
+  override def toString: String = {
+    s"""BlockHash {
+       |hash: ${Hex.toHexString(hash.toArray[Byte])}
+       |number: $number
+       |}""".stripMargin
+  }
+}
+
+object GetBlockHeaders {
+  implicit val rlpEndDec = new RLPEncoder[GetBlockHeaders] with RLPDecoder[GetBlockHeaders] {
+    override def encode(obj: GetBlockHeaders): RLPEncodeable = {
+      import obj._
+      block match {
+        case Left(blockNumber) => RLPList(blockNumber, maxHeaders, skip, reverse)
+        case Right(blockHash) => RLPList(blockHash.toArray[Byte], maxHeaders, skip, reverse)
+      }
+    }
+
+    override def decode(rlp: RLPEncodeable): GetBlockHeaders = rlp match {
+      case RLPList((block: RLPValue), maxHeaders, skip, reverse) if block.bytes.length < 32 =>
+        GetBlockHeaders(Left(block), maxHeaders, skip, reverse)
+
+      case RLPList((block: RLPValue), maxHeaders, skip, reverse) =>
+        GetBlockHeaders(Right(ByteString(block: Array[Byte])), maxHeaders, skip, reverse)
+
+      case _ => throw new RuntimeException("Cannot decode GetBlockHeaders")
+    }
+  }
+
+  val code: Int = 0x10 + 0x03
+}
+
+case class GetBlockHeaders(block: Either[BigInt, ByteString], maxHeaders: BigInt, skip: BigInt, reverse: Int) extends Message {
```

Comment:
is it `Either[BigInt, ByteString]` because you want to give more options to the caller? Maybe we can just stick to one of these?